makeflow/src/makeflow_mpi_starter.c

```diff
@@ -0,0 +1,336 @@
+    char procname[MPI_MAX_PROCESSOR_NAME];
+    int procnamelen;
+    MPI_Get_processor_name(procname, &procnamelen);
+    int rank_0_cores = 1;
+
+    struct hash_table* comps = hash_table_create(0, 0);
+    struct hash_table* sizes = hash_table_create(0, 0);
+
+    if (mpi_rank == 0) { //Master to decide who stays and who doesn't
+        int i;
+
+        for (i = 1; i < mpi_world_size; i++) {
+            unsigned len = 0;
+            MPI_Recv(&len, 1, MPI_UNSIGNED, i, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+            char* str = malloc(sizeof (char*)*len + 1);
+            memset(str, '\0', sizeof (char)*len + 1);
+            MPI_Recv(str, len, MPI_CHAR, i, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+
+            struct jx* recobj = jx_parse_string(str);
+            char* name = (char*) jx_lookup_string(recobj, "name");
+            int rank = jx_lookup_integer(recobj, "rank");
+
+
+            if (hash_table_lookup(comps, name) == NULL) {
+                long long p = rank;
+                hash_table_insert(comps, name, (void*) p);
+            }
+            //for partition sizing
+            if (hash_table_lookup(sizes, name) == NULL) {
+                long long p = 1;
+                hash_table_insert(sizes, name, (void*) p);
+            } else {
+                long long val = (long long) hash_table_lookup(sizes, name);
+                hash_table_remove(sizes, name);
+                long long p = val + 1;
+                hash_table_insert(sizes, name, (void*) p);
+
+            }
+
+            jx_delete(recobj);
+
+        }
+        for (i = 1; i < mpi_world_size; i++) {
+            hash_table_firstkey(comps);
+            char* key;
+            long long value;
+            int sent = 0;
+            while (hash_table_nextkey(comps, &key, (void**) &value)) {
+                if (value == i) {
+                    fprintf(stderr, "Telling %lli to live\n", value);
+                    MPI_Send("LIVE", 4, MPI_CHAR, i, 0, MPI_COMM_WORLD);
+                    sent = 1;
+                }
+            }
+            if (sent == 0) {
+                MPI_Send("DIE ", 4, MPI_CHAR, i, 0, MPI_COMM_WORLD);
+            }
+        }
+
+    } else { //send proc name and num
+        hash_table_delete(comps);
+        char* sendstr = string_format("{\"name\":\"%s\",\"rank\":%i}", procname, mpi_rank);
+        unsigned len = strlen(sendstr);
+        MPI_Send(&len, 1, MPI_UNSIGNED, 0, 0, MPI_COMM_WORLD);
+        MPI_Send(sendstr, len, MPI_CHAR, 0, 0, MPI_COMM_WORLD);
+
+        free(sendstr);
+        //get if live or die
+        char livedie[10];
+        MPI_Recv(livedie, 4, MPI_CHAR, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        if (strstr(livedie, "DIE")) {
+            MPI_Finalize();
+            return 0;
+        } else if (strstr(livedie, "LIVE")) {
+            //do nothing, continue
+        } else {
+            fprintf(stderr, "livedie string got corrupted, wrong command sent.... %s\n", livedie);
```

Comment:
Please remove from mpi_starter all the debugging fprintf(stderr,...

Also, makeflow_mpi_starter always seem to have debug output on. (As if -d is always specified.)

Reply:
fixed